src/sage/combinat/permutation.py

```diff
@@ -5382,154 +5386,139 @@ def rewind(L, k):
+            sage: Sigma.has_nth_root(3)
             True
             
             sage: Sigma = Permutation('(1, 3)')
-            sage: Sigma.has_kth_root()
+            sage: Sigma.has_nth_root(2)
             False
         
         .. SEEALSO::
         
-                * :meth:`kth_roots`
-                * :meth:`number_of_kth_roots`
+                * :meth:`nth_roots`
+                * :meth:`number_of_nth_roots`
         
         TESTS:
             
         We compute the number of permutations that have square roots (i.e. squares in `S_n`, :oeis:`A003483`)::
             
-            sage: [len([p for p in Permutations(n) if p.has_kth_root()]) for n in range(2, 7)]
+            sage: [len([p for p in Permutations(n) if p.has_nth_root(2)]) for n in range(2, 7)]
             [1, 3, 12, 60, 270]
             
-            sage: Permutation('(1)').has_kth_root()
+            sage: Permutation('(1)').has_nth_root(2)
+            True
+
+            sage: Permutation('').has_nth_root(2)
+            True
+
+            sage: Sigma = Permutations(6).random_element()
+            sage: Sigma.has_nth_root(1)
             True
             
-            sage: Permutations(4).identity().has_kth_root(-1)
+            sage: Permutations(4).identity().has_nth_root(-1)
             Traceback (most recent call last):
             ...
-            ValueError: k must be at least 1
+            ValueError: n must be at least 1
         """
 
         from sage.combinat.partition import Partitions
         from sage.arith.misc import divisors, gcd
+        from sage.rings.integer import Integer
 
-        if k < 1:
-            raise ValueError('k must be at least 1')
-        
-        P = Permutations(self.size())
+        if n < 1:
+            raise ValueError('n must be at least 1')
         
-        # Creating dict {length: number of cycles of this length in the cycle decomposition of self}
-        Cycles = {}
-        for c in self.cycle_tuples(singletons=True):
-            lc = len(c)
-            if lc not in Cycles:
-                Cycles[lc] = 0
-            Cycles[lc] += 1
-        
-        # for each length m, check if the number of m-cycles can come from a k-th power (i.e. if you can partitionate m*Cycles[m] into parts of size l with l = m*gcd(l,k))
-        for m in Cycles:
-            N = Cycles[m]
-            parts = [x for x in divisors(k) if gcd(m*x, k) == x]
-            if not len(Partitions(N, parts_in=parts)):
+        Cycles = self.cycle_type().to_exp_dict()
+
+        # for each length m, check if the number of m-cycles can come from a n-th power
+        # (i.e. if you can partitionate m*Cycles[m] into parts of size l with l = m*gcd(l, n))
+        for m, N in Cycles.items():
+            N = Integer(N) # I don't know why but ._findfirst doesn't work without
+            parts = [x for x in divisors(n) if gcd(m*x, n) == x]
+            if not Partitions(0, parts_in=[])._findfirst(N, parts):
```

Comment:
This is a really strange line of code.  `Partitions(0, parts_in=[])` is simply the empty partition.  `_findfirst` is a method without documentation, what is it suppose to do?

Reply:
My goal here is just to know if there exists a partition of `N` with parts in `parts`. I haven't been able to find an efficient method to do it.
Here, I create a parent Partitions_parts_in (writing `Partitions_parts_in(N, parts)` does not seem to work), and then ask for the first element.
If you have a better idea, I'll take it!

Reply:
`Partitions(N, parts_in=parts).is_empty()`

Reply:
Thanks.